stdlib/LinearAlgebra/src/matmul.jl

```diff
@@ -417,10 +458,15 @@ end
 end
 
 @inline function mul!(C::StridedMatrix{T}, adjA::Adjoint{<:Any,<:StridedVecOrMat{T}}, adjB::Adjoint{<:Any,<:StridedVecOrMat{T}},
-                 alpha::Union{T, Bool}, beta::Union{T, Bool}) where {T<:BlasFloat}
+                 α::Number, β::Number) where {T<:BlasFloat}
     A = adjA.parent
     B = adjB.parent
-    return gemm_wrapper!(C, 'C', 'C', A, B, MulAddMul(alpha, beta))
+    alpha, beta = promote(α, β, zero(T))
+    if alpha isa T && beta isa T
+        return gemm_wrapper!(C, 'C', 'C', A, B, MulAddMul(alpha, beta))
+    else
+        return generic_matmatmul!(C, 'C', 'C', A, B, MulAddMul(α, β))
+    end
```

Comment:
Looking at many repetitions like this, I started to wonder my approach was wrong.  Maybe instead we should put the branch in `gemm_wrapper!`?

Reply:
Yea, this was a very naive approach, but it helped me understand the architecture better. I think I'll revert this and start over again. Do you think we need `MulAddMul` in the `gemv` and `gemm` section at all? I don't see any dispatch on the parameters of `MulAddMul`. Also, I think we need to avoid promoting `Bool` values.

Reply:
Yeah, it would be great if we can pass alpha and beta to `gemm_wrapper!` as-is and construct `MulAddMul` just before invoking `generic_matmatmul!` via `gemm_wrapper!`.  Then there would be no construction of `MulAddMul` when calling BLAS.  Of course, we need to rely on the compiler to propagate the constants much longer...